popde/adaptive_kde.py

```diff
@@ -224,6 +224,151 @@ def optimize_parameters(self, method='loo_cv', fom_plot_name=None):
+            backend (str, optional): Backend to use, default is 'KDEpy'.
+            bandwidth (float, optional): Fixed bandwidth value, default is 1.0.
+            alpha (float, optional): Initial alpha value, default is 0.0.
+            dim_names (list, optional): Dimension names.
+            do_fit (bool, optional): Whether to fit the KDE upon initialization.
+            n_splits (int, optional): Number of splits for k-fold 
+               cross-validation.
+        """
+
+        self.n_splits = n_splits
+        self.alpha = alpha
+        self.bandwidth = bandwidth
+
+        super().__init__(data, weights, input_transf, stdize, rescale, backend, 
+                bandwidth, alpha, dim_names, do_fit)
+
+    def loo_cv_score(self, rescale_factors_alpha):
+        """
+        Compute the Leave-One-Out Cross-Validation (LOO-CV) score
+          based on log-likelihood.
+
+        Args:
+            rescale_factors_alpha (array-like): Array of rescale factors per
+              dimension and alpha value.
+
+        Returns:
+            float: Negative sum of log-likelihood scores for LOO-CV.
+        """
+        rescale_factors = rescale_factors_alpha[:-1]
+        alpha_choice = rescale_factors_alpha[-1]
+        from sklearn.model_selection import LeaveOneOut
+        loo = LeaveOneOut() 
+        fom = 0.
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factors,
+                                  bandwidth=self.bandwidth, alpha=alpha_choice)
+            fom += np.log(awkde.evaluate_with_transf(test_data))
+        return -fom
+
+    def kfold_cv_score(self, rescale_factors_alpha, seed=42):
+        """
+        Perform k-fold cross-validation
+        """
+        rescale_factors = rescale_factors_alpha[:-1]
+        alpha_choice = rescale_factors_alpha[-1]
+        from sklearn.model_selection import KFold
+        kf = KFold(n_splits=self.n_splits, shuffle=True, random_state=seed)
+        fom = []
+        for train_index, test_index in kf.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None # FIX ME
+
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factors,
+                                  bandwidth=self.bandwidth, alpha=alpha_choice, do_fit=True)
+
+            log_kde_eval = np.log(awkde.evaluate_with_transf(test_data))
+            fom.append(log_kde_eval.sum())
+        return -sum(fom)
+
+    def optimize_rescale_parameters(self, initial_rescale_factor, initial_alpha=0.0, method='kfold_cv', bounds=None):
```

Comment:
I think we prefer initial alpha to be nonzero since it may not be good for the optimization if we initialize with a value at the exact boundary.

Reply:
Also, in this method we should have `**kwargs` for additional options that we pass on to the minimize function (like tolerance, printing progress etc.)

Reply:
Also, we might want to control both the cross validation method and the optimization method.  So use eg `cv_method=` and `opt_method=` kwargs

Reply:
done